crates/autopilot/src/decoded_settlement.rs

```diff
@@ -0,0 +1,488 @@
+}
+
+impl From<(Address, U256, Bytes<Vec<u8>>)> for DecodedInteraction {
+    fn from((target, value, call_data): (Address, U256, Bytes<Vec<u8>>)) -> Self {
+        Self {
+            target,
+            value,
+            call_data,
+        }
+    }
+}
+
+impl From<DecodedSettlementTokenized> for DecodedSettlement {
+    fn from((tokens, clearing_prices, trades, interactions): DecodedSettlementTokenized) -> Self {
+        DecodedSettlement {
+            tokens,
+            clearing_prices,
+            trades: trades
+                .into_iter()
+                .map(|trade| DecodedTrade {
+                    sell_token_index: trade.0,
+                    buy_token_index: trade.1,
+                    receiver: trade.2,
+                    sell_amount: trade.3,
+                    buy_amount: trade.4,
+                    valid_to: trade.5,
+                    app_data: trade.6,
+                    fee_amount: trade.7,
+                    flags: trade.8,
+                    executed_amount: trade.9,
+                    signature: trade.10,
+                })
+                .collect(),
+            interactions: interactions.map(|inner| inner.into_iter().map(Into::into).collect()),
+        }
+    }
+}
+
+pub struct FeeConfiguration {
+    pub fee_objective_scaling_factor: f64,
+}
+
+pub struct Order {
+    pub full_fee_amount: U256,
+    pub kind: OrderKind,
+    pub sell_token: H160,
+    pub sell_amount: U256,
+    pub buy_amount: U256,
+    pub signature: Vec<u8>, //encoded signature
+}
+
+impl From<model::order::Order> for Order {
+    fn from(order: model::order::Order) -> Self {
+        Self {
+            full_fee_amount: order.metadata.full_fee_amount,
+            kind: order.data.kind,
+            sell_token: order.data.sell_token,
+            sell_amount: order.data.sell_amount,
+            buy_amount: order.data.buy_amount,
+            signature: order
+                .signature
+                .encode_for_settlement(order.metadata.owner)
+                .to_vec(),
+        }
+    }
+}
+
+impl DecodedSettlement {
+    pub fn new(input: &[u8]) -> Result<Self> {
+        let function = GPv2Settlement::raw_contract()
+            .abi
+            .function("settle")
+            .unwrap();
+        let decoded_input = decode_function_input(function, input)?;
+        <DecodedSettlementTokenized>::from_token(Token::Tuple(decoded_input))
+            .map(Into::into)
+            .context("failed to decode settlement")
+    }
+
+    /// Returns the total surplus denominated in the native asset for the
+    /// solution.
+    pub fn total_surplus(&self, external_prices: &ExternalPrices) -> U256 {
+        self.trades.iter().fold(0.into(), |acc, trade| {
+            acc + match surplus(trade, &self.tokens, &self.clearing_prices, external_prices) {
+                Some(surplus) => surplus,
+                None => {
+                    tracing::warn!("possible incomplete surplus calculation");
```

Comment:
I'm unsure about this warn. If this is something that shouldn't happen for valid settlements, then this function should return Result. If it is expected to happen then there shouldn't be a warning. There are two similar cases in `total_fees`.